src/libopensc/pkcs15-cardos.c

```diff
@@ -0,0 +1,177 @@
+			switch(sa->mechanism) {
+			case 0x01 :
+				/*
+				 * Card appears to use lower 4 bits of reference as key, and upper
+				 * 4 bits as mech for card.
+				 * Also has a bug if mechanism = 1 (CKM_RSA_PKCS1) and reference 0x10 
+				 * bit is set mechanism should be 3 (CKM_RSA_X_509) 
+				 * correct the mechanism in tokenInfo
+				 */
+				if (sa->reference & 0x10) {
+					sc_log(card->ctx, "Changeing mechanism to CKM_RSA_X_509 based on reference");
+					passed->new_flags |= SC_ALGORITHM_RSA_RAW
+						| SC_ALGORITHM_RSA_PAD_NONE;
+					sa->mechanism = 0x03;
+				} else
+				passed->new_flags |= SC_ALGORITHM_RSA_PAD_PKCS1;
+				break;
+			case 0x03 :
+				passed->new_flags |= SC_ALGORITHM_RSA_RAW
+					| SC_ALGORITHM_RSA_PAD_NONE;
+				break;
+			case 0x06 :
+				passed->new_flags |= SC_ALGORITHM_RSA_HASH_SHA1;
+				break;
+			case 0x1041:
+				passed->ec_flags |= SC_ALGORITHM_ECDSA_RAW;
+				/* no old_ec_flags */
+				/* TODO turn on sizes from  ec curves OIDS */
+				break;
+			default:
+				sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "UNKNOWN MECH: 0x%8.8x", sa->mechanism);
+			}
+
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "New_flags 0x%8.8lx New_ec_flags: 0x%8.8lx",
+				passed->new_flags, passed->ec_flags);
+		}
+
+		if (passed->new_flags == 0) {
+			if (p15card->tokeninfo && p15card->tokeninfo->flags & SC_PKCS15_TOKEN_EID_COMPLIANT) {
+				sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "EID_COMPLIANT flag found");
+				passed->new_flags = (passed->used_flags & ~SC_ALGORITHM_SPECIFIC_FLAGS) | SC_ALGORITHM_RSA_PAD_PKCS1;
+			} else
+				passed->new_flags = passed->used_flags; /* from default cardos_init */
+		}
+	}
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL,"Final New_flags 0x%8.8lx New_ec_flags: 0x%8.8lx", passed->new_flags, passed->ec_flags);
+
+	passed->pass = 2; /* tell card driver to use the new flags */
+	r = sc_card_ctl(p15card->card, SC_CARDCTL_CARDOS_PASS_ALGO_FLAGS, passed);
+
+	free(passed);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+static int
+cardos_pkcs15emu_detect_card(sc_pkcs15_card_t *p15card)
+{
+	if (p15card->card->type <  SC_CARD_TYPE_CARDOS_BASE)
+		return SC_ERROR_WRONG_CARD;
+
+	if (p15card->card->type >= SC_CARD_TYPE_CARDOS_BASE + 1000)
+		return SC_ERROR_WRONG_CARD;
+
+	return SC_SUCCESS;
+}
+
+
+static int
+sc_pkcs15emu_cardos_init(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
+{
+	sc_card_t *card = p15card->card;
+	int r;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	r = sc_pkcs15_bind_internal(p15card, aid);
+	LOG_TEST_RET(card->ctx, r, "sc_pkcs15_bind_internal failed");
+
+	/* If card has created algorithms, return  */
+	sc_log(card->ctx, " card->algorithms:%p card->algorithm_count:%d", card->algorithms, card->algorithm_count);
+	if (!card->algorithms && card->algorithm_count == 0) {
+		r = cardos_fix_token_info(p15card);
+	}
+
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+
+int
+sc_pkcs15emu_cardos_init_ex(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
+{
```

Comment:
clang-format suggested the following change:


```
-
-int
-sc_pkcs15emu_cardos_init_ex(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
-{
+int sc_pkcs15emu_cardos_init_ex(struct sc_pkcs15_card *p15card, struct sc_aid *aid) {
```

<!-- identifier: clang-format -->
<!-- File: src/libopensc/pkcs15-cardos.c -->